app/helpers/govspeak_helper.rb

```diff
@@ -243,39 +250,40 @@ def extract_number_from_heading(nokogiri_el)
     nokogiri_el.inner_text[/^\d+.?[^\s]*/]
   end
 
-  def markup_to_nokogiri_doc(govspeak, images = [], allowed_elements = [])
-    govspeak = build_govspeak_document(govspeak, images, allowed_elements)
+  def markup_to_nokogiri_doc(govspeak, images = [], attachments = [])
```

Comment:
That's correct. The transformation is a little muddled, but broadly speaking it works like this:

```mermaid
flowchart TD
    A["User input"]
    B["Whitehall\npre-processors"]
    C["Govspeak\npre-processors"]
    D["Kramdown gem\nconverts Markdown to HTML"]
    E["Govspeak\npost-processors"]
    F["Whitehall post-processors"]
    H["HTML output"]
    A --> |"Markdown\n(may also contain 'allowed' HTML)"| B
    B --> |Markdown with 'allowed' HTML| C
    C --> |Markdown with 'allowed' HTML| D
    D --> |Sanitized HTML| E
    E --> |HTML| F
    F --> |HTML| H
```

Before this PR, the attachment embedding happened in a **Whitehall pre-processor**. Therefore the HTML output from that embed process needed to be considered 'safe'. [In some circumstances](https://github.com/alphagov/whitehall/blob/9edcdce940320149eb10b2d4d80559b12083925d/app/views/documents/_attachment.html.erb#L77-L95) the attachment HTML contained a `<details>` element, which isn't in the [default list of allowed elements](https://github.com/alphagov/govspeak/blob/54fb05496f62b0695f2d5b79a1cff96830fdf0b1/lib/govspeak/html_sanitizer.rb#L44-L69).

After this PR, the attachment embedding happens in a **Govspeak post-processor**. So the HTML output from that process can contain anything – the sanitisation has already happened, so any HTML elements are allowed at this point.

This is why we don't need to specify any `allowed_elements` any more – because the attachment now gets embedded after sanitisation.

Reply:
Thanks for the great explanation 😄